core/src/main/java/org/apache/iceberg/types/PruneUnknownTypes.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.types;
+
+import java.util.List;
+import java.util.stream.IntStream;
+import org.apache.hadoop.util.Lists;
+import org.apache.iceberg.Schema;
+
+/** This is used to remove UnknownTypes that should not be persistent in files */
+public class PruneUnknownTypes extends TypeUtil.SchemaVisitor<Type> {
+  private static final PruneUnknownTypes INSTANCE = new PruneUnknownTypes();
+
+  /**
+   * Prunes any UnknownType from a Schema
+   *
+   * @param schema a Schema
+   */
+  public static Schema convert(Schema schema) {
+    Types.StructType struct = (Types.StructType) TypeUtil.visit(schema, INSTANCE);
+    return new Schema(struct.fields(), schema.identifierFieldIds());
+  }
+
+  /**
+   * Visits a schema and removes the UnknownTypes.
+   *
+   * <p>Used when creating a write schema, and skip over the UnknownTypes, since they are not
+   * materialized in the data files
+   */
+  private PruneUnknownTypes() {}
+
+  public static Types.StructType convert(Types.StructType structType) {
+    Object obj = TypeUtil.visit(structType, INSTANCE);
+
+    if (obj instanceof Types.StructType) {
+      return (Types.StructType) obj;
+    } else {
+      return ((Schema) obj).asStruct();
+    }
+  }
+
+  @Override
+  public Type schema(Schema schema, Type structResult) {
+    if (structResult.typeId().equals(Type.TypeID.UNKNOWN)) {
+      return Types.StructType.of();
+    } else {
+      return structResult;
+    }
+  }
+
+  @Override
+  public Type struct(Types.StructType struct, List<Type> fieldResults) {
+    List<Types.NestedField> fields = struct.fields();
+    boolean needsRewrite =
```

Comment:
I find this to be hard to read compared to a private method that returns when it hits a match.

Also, I would probably just combine this with the next loop to produce the new fields. That avoids two loops that run the same expensive checks (`Type#equals`) and the only cost is adding unchanged fields to the `newFields` list and discarding it if the struct doesn't change.